frame/nomination-pools/src/migration.rs

```diff
@@ -437,8 +437,8 @@ pub mod v3 {
 		#[cfg(feature = "try-runtime")]
 		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
 			ensure!(
-				Pallet::<T>::current_storage_version() > Pallet::<T>::on_chain_storage_version(),
-				"the on_chain version is equal or more than the current one"
+				Pallet::<T>::current_storage_version() >= Pallet::<T>::on_chain_storage_version(),
+				"nomination-pools::migration::v3: on_chain version is greater than current version"
```

Comment:
Doesn't it prevent someone from accidentally applying a old migration?

Reply:
> If we want to keep them in the runtime, then they should not magically start failing after they are applied on chain.

I completely agree here. There're multiple instances where that happens, I think we need some guidelines written up mentioning this.

In this particular case, wouldn't it help to prevent someone from accidentally upgrading with an old pallet version? Or maybe that is so unlikely we shouldn't bother checking for it?

Reply:
Yes we need https://github.com/paritytech/substrate/issues/13107

Reply:
Actually no they are useless.. Because they'll be skipped in the actual `on_runtime_upgrade`. Removing them.

Reply:
Quick question about the nom pool v3 migration, it's checked using `if current > onchain` https://github.com/paritytech/substrate/blob/e85188b800c63ca3be545c1a43a0927fe0b6fd11/frame/nomination-pools/src/migration.rs#L411

This seems not right, should I update it to explicitly check `on_chain == 2`?

Reply:
> Yes we need #13107

I think #13107 will apply only within the `fn on_runtime_upgrade()` hook, not the pre/post hooks?

Reply:
> This seems not right, should I update it to explicitly check current == 3 && on_chain == 2 

Yes please.

Reply:
Thanks @bkchr, I've updated that migration